man/fi_profile.3.md

```diff
@@ -0,0 +1,260 @@
+
+# VARIABLES
+
+The profiling variables are defined in providers, and become accessable 
+(readable) externally through the profiling interface.  A variable can be 
+a configuration setting, a runtime state or a counter in an endpoint or a 
+domain in a provider. Some variables are common and available in most 
+providers, while others are provider-specific, and are only available in 
+one provider. Users can query available variables and their values using 
+the profiling interface.
+
+FI_VAR_UNEXP_MSG_CNT is a common variable and it counts the number of  
+unexpected message recevied and unmatched. The variable has a unint64_t 
+data value.
+
+A variable is specified using struct fi_profile_desc and is described below.
+
+# EVENTS
+
+The profiling event is defined by the providers. It is used to notify users 
+that an operation has occured or a state has been moved, which may change 
+the values of some variables. Event is a way for providers to inform changes 
+on the variables. Similar to the variables, some events are pre-defined and 
+shared by many providers, while others are provider-specific events and 
+defined by each provider. Users can register a callback function on an event 
+to get notified when the event occurs.
+
+FI_EVENT_UNEXP_MSG_RECVD and FI_EVENT_UNEXP_MSG_MATCHED are common events 
+which imply there are changes on variable FI_VAR_UNEXP_MSG_CNT.
+
+Same as variables, events are specified by the fi_profile_desc struct.  
+
+{% highlight c %}
+struct fi_profile_desc {
+	uint32_t id;
+	enum fi_datatype type;
+	size_t size;
+	const char *name;
+	const char *desc;
+};
+{% endhighlight %}
+
+## id
+
+Each variable or event has a unique identifier (id). Providers provide this 
+id and users use it to get the variable value or register callbacks for 
+the event.
+
+## type
+
+The data type of the variable. 
+
+## size
+
+The size (in byte) of the data. "type" and "size" can be used to support 
+any provider-specific data.
+
+## name
+
+Reference to the name of a variable or an event. It is either defined or 
+allocated by providers. Variables or events with the same name should have 
+an identical id. 
+
+## desc
+Reference to the description of a variable or an event. It is either defined 
+or allocated by providers.
+
+# CALLBACK
+
+Callback functions are used to receive notifications from a provider when 
+an event occurs. A callback function is defined by the user and called by 
+a provider as an additional operation, normally before or after data transfer. 
+This will increase the overhead of the operation.  Callback functions should 
+avoid having blocking calls, or taking non-trival time. And callback functions 
+should limit interactions with a provider to reading the variables only.
+
+The signature of the callabck function is:
+
+{% highlight c %}
+int (*callback)(struct fid_profile *prof_fid, uint32_t event,
+                void *param, size_t size, void *context)
+{% endhighlight %}
+
+
+## fi_profile_open
```

Comment:
Set to '#' to set function call as top-level item.  Apply this to all functions defined below.  E.g.
```
# fi_profile_open
open blah blah....

# fi_profile_close
close blah blah
```